docs/testing-with-django.md

```diff
@@ -1,8 +1,237 @@
+in `zerver/tests` to get a feel for the patterns we use.
+
+A good practice is to get a "failing test" before you start to implement
+your feature.  First, it is a useful exercise to understand what needs to happen
+in your tests before you write the code, as it can help drive out simple
+design or help you make incremental progress on a large feature.  Second,
+you want to avoid introducing tests that give false positives.  Ensuring
+that a test fails before you implement the feature ensures that if somebody
+accidentally regresses the feature in the future, the test will catch
+the regression.
+
+Another important file to skim is
+[zerver/lib/test_helpers.py](https://github.com/zulip/zulip/blob/master/zerver/lib/test_helpers.py),
+which contains test helpers and our `AuthedTestCase` class.
+
+### Setting up data for tests
+
+All tests start with the same fixture data.  (The tests themselves
+update the database, but they do so inside a transaction that gets
+rolled back after the tests complete.)
+
+The fixture data includes a few users that are named after
+Shakesepeare characters, and they are part of the "zulip.com" realm.
+
+Generally, you will also do some explicit data setup of your own. Here
+are a couple useful methods in AuthedTestCase:
+
+- common_subscribe_to_streams
+- send_message
+- subscribe_to_stream
+
+More typically, you will use methods directly from the backend code.
+(This ensures more end-to-end testing, and avoids false positives from
+tests that might not consider ancillary parts of data setup that could
+influence tests results.)
+
+Here are some example action methods that tests may use for data setup:
+
+- check_send_message
+- create_stream_if_needed
+- do_add_subscription
+- do_change_is_admin
+- do_create_user
+- do_make_stream_private
+
+## Zulip Testing Philosophy
+
+If there is one word to describe Zulip's philosophy for writing tests,
+it is probably "flexible."  (Hopefully "thorough" goes without saying.)
+
+When in doubt, unless speed concerns are prohibitive,
+you usually want your tests to be somewhat end-to-end, particularly
+for testing endpoints.
+
+These are some of the testing strategies that you will see in the Zulip
+test suite...
+
+### Endpoint tests
+
+We strive to test all of our URL endpoints.  The vast majority of Zulip
+endpoints support a JSON interface.  Regardless of the interface, an
+endpoint test generally follows this pattern:
+
+- Set up the data.
+- Login with `self.login()` or set up an API key.
+- Use a Zulip test helper to hit the endpoint.
+- Assert that the result was either a success or failure.
+- Check the data that comes back from the endpoint.
+
+Generally, if you are doing endpoint tests, you will want to create a
+test class that is a subclass of `AuthedTestCase`, which will provide
+you helper methods like the following:
+
+- api_auth
+- assert_json_error
+- assert_json_success
+- client_get
+- client_post
+- get_api_key
+- get_streams
+- login
+- send_message
```

Comment:
I think the main area I'd want to expand in this page docs is adding some more detail around this sublist.  Definitely happy to merge as-is and have that be a followup project, though.